packages/react-compose/src/createOptionsResolver.ts

```diff
@@ -0,0 +1,67 @@
+import { ComposePreparedOptions, GenericDictionary } from './types';
+import { getNativeElementProps } from '@uifabric/utilities';
+import { resolveClasses } from './resolveClasses';
+import { resolveSlotProps } from './resolveSlotProps';
+
+export type OptionsResolverResult = {
+  state: GenericDictionary;
+  slots: GenericDictionary;
+  slotProps: GenericDictionary;
+};
+
+/**
+ * A set of mapped props for intrinsic element types.
+ */
+export const defaultMappedProps: Record<string, string> = {
+  iframe: 'src',
+  img: 'src',
+  input: 'type',
+};
+
+export const EmptyRender = () => null;
+
+/**
+ * Creates an options resolve function which should attach to an options object
+ * for a composed component.
+ */
+export const createOptionsResolver = <TState>(options: ComposePreparedOptions) => {
+  // Returning a function so that "resolve" function shows up in profiling.
+  // tslint:disable-next-line:no-function-expression
+  return function resolve(stateOrProps: TState): OptionsResolverResult {
+    const slotProps: OptionsResolverResult['slotProps'] = {};
+    const state = stateOrProps as OptionsResolverResult['state'];
+    const slots: OptionsResolverResult['slots'] = { ...options.slots };
+    const result: OptionsResolverResult = {
+      state,
+      slotProps,
+      slots,
+    };
+
+    // Always ensure a root slot exists.
+    slots.root = state.as || slots.root || 'div';
```

Comment:
Shouldn't this go before the assignment to `result` in line 34?

Reply:
it won't make a difference, but maybe it cleans up the code. The `result` assignment will assign the `slots` reference. We're mutating values but that won't change the reference.